exercises/01-react-stateful/src/modules/root/components/header.tsx

```diff
@@ -0,0 +1,9 @@
+import React from 'react'
+
+interface HeaderProps {
+  title: string
```

Comment:
In my current project, we have convention to add `readonly` before every property (and before every array as well) to prepare as immutable props as possible.

There's also `Readonly` interface that converts all fields to `readonly` variants but it converts only the first level. Let's say we have a property like `users: User[]` here as well. Using the `Readonly` interface would change

```ts
interface HeaderProps {
  title: string;
  users: User[];
}
```

to

```ts
interface HeaderProps {
  readonly title: string;
  readonly users: User[];
}
```

but we want something like

```ts
interface HeaderProps {
  readonly title: string;
  readonly users: readonly User[];
}
```

and I'm not aware of any built-in way.

I would personally prepare (or install if there's a lib) something like `DeepReadonly` and use

```ts
const Header: React.FC<DeepReadonly<HeaderProps>> = ...
```

but that's too complicated for beginners IMHO. Anyway, I think that we should mention that it's not a good idea to modify input arguments and we can prevent it with `readonly`. Therefore, I would suggest to use `readonly title: string` here. Other opinions? :slightly_smiling_face:

Reply:
I've found there's an implementation of `DeepReadonly` in [ts-essentials](https://github.com/krzkaczor/ts-essentials), which seems to work as you expect.

Reply:
There's also a [user-types.ts](https://github.com/salsita/react-training/blob/43d1ccace36c096a0d34498491f91019aca9f9fc/exercises/01-react-stateful/src/modules/users/user-types.ts) file, which I was going to rewrite as:
```ts
export interface UserData {
  firstName: string
  lastName: string
}

export interface User extends UserData {
  id: number
}

```

And use like this in [user-list.tsx](https://github.com/salsita/react-training/blob/43d1ccace36c096a0d34498491f91019aca9f9fc/exercises/01-react-stateful/src/modules/users/components/user-list.tsx):
```ts
const addUser = (user: UserData) => {
    setUsers((prevUsers) => [
      ...prevUsers,
      { id: prevUsers.length + 1, ...user },
    ])
  }
```
instead of 
```ts
const addUser = (user: { firstName: string; lastName: string }) => {/*...*/}
```

1. Would it be an acceptable approach?
2. Would this readonly rule apply also for `User` and `UserData`?
3. How do you normally approach the situation when you want to use a mutable interface as a parameter type to a function which must not edit this interface? `Readonly` seems as a good solution.

Reply:
1. `UserData` + `User` - yeah, that's fine. :+1: 
2. I wouldn't use `readonly` in those interfaces directly.
3. This is a bit tricky and depends on how strict we want to be. We should also keep in mind that the training project (especially this first exercise) shouldn't be very complicated. Like I post above, in a real project where I would like to be strict, I would use `const Header: React.FC<DeepReadonly<HeaderProps>> = ...` but it might be too complicated/magical for the first exercise.

Reply:
I propose the following changes:
1. Add `User` and `UserData` interfaces
2. Add `readonly` to `HeaderProps` interface
3. Update README. Please check whether the description of user types is acceptable for tutorial -- it might be too detailed.

Changes are made in several commits, see [combined diff](
https://github.com/salsita/react-training/pull/44/files/43d1ccace36c096a0d34498491f91019aca9f9fc..9ffcf4e6319961e0799e4299d57c968da2df92bb)

Thank you!